### PR TITLE
Construct transform on init of sliceinfo and remove circular dependence in sliceviewer

### LINKED ISSUE #32436: Sliceviewer out of range error when non-orthog transform a 4D MD workspace with non-Q dim first
**Describe the bug**
Having a 4D MD workspace with non-Q dim first seems to give an out-of-range error when applying a non-orthogonal transform - the error comes from the model which calculates the NonOrthogonalTransfomr from a  projection matrix which is always a 3 x 3 matrix and assumes any non-Q dimensions are after the Q dimensions.

```
Traceback (most recent call last):
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\view.py", line 429, in on_non_orthogonal_axes_toggle
    self.presenter.nonorthogonal_axes(state)
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\presenter.py", line 352, in nonorthogonal_axes
    self.model.create_nonorthogonal_transform(self.get_sliceinfo()))
  File "C:\mantid\qt\python\mantidqt\widgets\sliceviewer\model.py", line 273, in create_nonorthogonal_transform
    display_indices.pop(slice_info.z_index)
IndexError: pop index out of range
```

**To Reproduce**
1. Run this script to make a workspace with energy along the first axis
```
md_4D = CreateMDWorkspace(Dimensions=4, Extents=[-1,1,-1,1,-1,1,-10,10], Names="H,K,L,E", 
    Frames='HKL,HKL,HKL,General Frame',Units='r.l.u.,r.l.u.,r.l.u.,meV')
print([md_4D.getDimension(ii).getMDFrame().isQ() for ii in range(0,4)])
expt_info = CreateSampleWorkspace()
md_4D.addExperimentInfo(expt_info)
SetUB(md_4D, 1,1,2,90,90,120)
BinMD(InputWorkspace='md_4D', AxisAligned=False, BasisVector0='E,m,0,0,0,1', BasisVector1='QH,m,1,0,0,0', BasisVector2='K,m,0,1,0,0', BasisVector3='L,m,0,0,1,0', OutputExtents='-2,2,-1,1,-0.5,0.5,-0.2,0.2', OutputBins='21,21,21,21', NormalizeBasisVectors=False, OutputWorkspace='out')
```
2. View the workspace `out` in the sliceviewer (note the non-orthogonal view button is erroneously enabled for slices including energy - see issue #32314)
3. Swap display axes to non-orthogonal pair (x,y) = (H,K)
4. Click the non-orthog view and it give the above error

**Expected behavior**
Performs a non-orthog transform without error

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Description of work.**

The transform uses the angle between the display axes. Before the transform was constructed in the model which required the viewing axes/display indices to be passed in the sliceinfo object. In addition the sliceinfo (which is used to get the transform for the matplotlib stuff and in the peaksviewer) was initialised with a transform stored in the data_view (which wasn't being kept updated with the display axes as it was only updated from the sliceinfo when nonorthog/orthog axes were created when the non-orthog view button is clicked).

Here we construct transform with angle between axes when the sliceinfo is init (using angles calculated in model using OrientedLattice on init). The transform in the data_view is only updated from the sliceinfo when non-orthog axes are created (so no longer any circular dependence - the transform is not passed back from data_view to sliceinfo).

The enabling of the non-orthogonal view button on opening is now done by querying the sliceinfo object - as one needs to know the display indices.

Also for MDHisto workspaces it now gets the projection matrix from the basis vectors on the workspace rather than search for the W_MATRIX log (which doesn't exist causing it to assume an identity matrix)

**To test:**

(1) Repeat tests in PR #32392 (check that non-ortho button is enabled when expected for various 2D,3D and 4D workspaces)

- To check the transform of a 2D MD workspace (technically it's still 3D but you can't edit the slice in the third dimentsion), open e.g. `ws_nonortho_svrebinned` in sliceviewer from the above

(2) Follow instructions in #32436 and check that 
- slices with energy have the non-ortho view disabled
- when viewing HK plane that non-ortho view is enabled and the transform is OK
- that switching from e.g. (x,y)=(H,K) to (x,y)=(H,Energy) that non-ortho view is disabled

(3) Check transform is correct when UB replaced or deleted e.g.
```
ws = CreateMDWorkspace(Dimensions='3', Extents='-6,6,-4,4,-0.5,0.5',
                       Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                       Frames='HKL,HKL,HKL',
                       SplitInto='2', SplitThreshold='50')

expt_info = CreateSampleWorkspace()
ws.addExperimentInfo(expt_info)
SetUB(ws, 1,1,2,90,90,120)
# i) open ws in sliceviewer
# ii) click non-orthog transform view
# iii) Change UB to orthogonal (note the plot won't update automatically)
SetUB(ws, 1,1,2,90,90,90)
# iv) Toggle non-ortho view - it should be correct now
```
(4) Check that when the frame is changed such that the support for non-ortho transform is different then sliceviewer closes with error
`The property supports_nonorthogonal_axes is different on the new workspace.`
To do this e.g. 
```
ws = CreateMDWorkspace(Dimensions='3', Extents='-6,6,-4,4,-0.5,0.5',
                       Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                       Frames='HKL,HKL,HKL',
                       SplitInto='2', SplitThreshold='50')
FakeMDEventData(ws, EllipsoidParams='1e+04,0,1,0,1,0,0,0,1,0,0,0,1,0.2,0.6,0.3,1', RandomSeed='3873875')
expt_info = CreateSampleWorkspace()
ws.addExperimentInfo(expt_info)
SetUB(ws, 1,1,2,90,90,120)

# open ws in sliceviewer here

ws = CreateMDWorkspace(Dimensions='3', Extents='-6,6,-4,4,-0.5,0.5',
               Names='Q1,Q2,Q3', Units='Ang-1,Ang-1,Ang-1',
               Frames='QLab,QLab,QLab',
               SplitInto='2', SplitThreshold='50')    
```
(5) Check that peaks can be overlaid (open sliveviewer > three blobs button > select `peaks`)
```
ws = CreateMDWorkspace(Dimensions='3', Extents='-6,6,-4,4,-2,2',
                       Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                       Frames='HKL,HKL,HKL',
                       SplitInto='2', SplitThreshold='50')
expt_info = CreateSampleWorkspace()
ws.addExperimentInfo(expt_info)
SetUB(ws, 1,1,2,90,90,120)
FakeMDEventData(ws, EllipsoidParams='1e+04,0,1,1,1,0,0,0,1,0,0,0,1,0.02,0.06,0.03,1', RandomSeed='3873875')
# add peak
CreatePeaksWorkspace(InstrumentWorkspace='ws', NumberOfPeaks=0, OutputWorkspace='peaks')
CopySample(InputWorkspace='ws', OutputWorkspace='peaks', CopyName=False, CopyMaterial=False, CopyEnvironment=False, CopyShape=False)
p = mtd['peaks']
pk = p.createPeakHKL([0,1,1]) # axes aligned with viewing axes
p.addPeak(pk)

IntegratePeaksMD(InputWorkspace='ws', PeakRadius=0.6, BackgroundInnerRadius=0.6,
    BackgroundOuterRadius=1, PeaksWorkspace='peaks', OutputWorkspace='peaks_int_centroid_unfixed', 
    Ellipsoid=True, UseCentroid=False, FixMajorAxisLength = True, UseOnePercentBackgroundCorrection=False)
```

Try toggling non-ortho view on and off - note that the position of the peak and the ellipsoid shape of the integrated peak should agree with the data.

(6) Check that the add/delete peak functionality works (with non-ortho view on and off)
(7) Test with workspace with other frames - e.g. QLab using this workspace
```
ws = CreateMDWorkspace(Dimensions='3', Extents='-6,6,-4,4,-0.5,0.5',
               Names='Q1,Q2,Q3', Units='Ang-1,Ang-1,Ang-1',
               Frames='QLab,QLab,QLab',
               SplitInto='2', SplitThreshold='50')   
```
(8) Check that non-axis aligned slices can be viewed in sliceviewer correctly
```
ws = CreateMDWorkspace(Dimensions='3', Extents='-3,3,-3,3,-3,3',
                       Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                       Frames='HKL,HKL,HKL',
                       SplitInto='2', SplitThreshold='10')
expt_info = CreateSampleWorkspace()
ws.addExperimentInfo(expt_info)
SetUB(ws, 1,1,2,90,90,120)
BinMD(InputWorkspace=ws, AxisAligned=False,
    BasisVector0='[00L],r.l.u.,0,0,1', 
    BasisVector1='[HH0],r.l.u.,1,1,0',
    BasisVector2='[-HH0],r.l.u.,-1,1,0',
    OutputExtents='-4,4,-4,4,-0.25,0.25', 
    OutputBins='101,101,1', OutputWorkspace='ws_slice', NormalizeBasisVectors=False)
```
Check `ws_slice` has non-ortho view enabled bu when it is clicked that the grid is orthogonal (in this case 110 is orthogonal to 001).

Fixes #32436 and #32469  (and permanent fix for #32314 - for which a temp. fix was done in PR #32392)

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by martyngigg — CHANGES_REQUESTED** [review]
Sorry it has taken me so long to get to this.

I think it looks good and I've just a couple of comments for code clarity

**Review by DanielMurphy22 — CHANGES_REQUESTED** [review]
I've run through the testing instructions above too and they all worked well for me. The script in section 5 gave a MaskBTP error which is expected.

I really like all of these instructions and I think it would be great to integrate some of them into a non-orthogonal view section in the SliceViewer Manual Testing instructions. I particularly like checking that non-orthogonal view self-disables when an energy axis is set, that the sliceviewer closes with the warning `property supports_nonorthogonal_axes is different...` when the ws changes, and that orthogonal axes with a valid UB are handled correctly. So I think that means the bits I'd like included in the Manual testing are 2) 4) and 8).

File: dev-docs/source/Testing/SliceViewer/SliceViewer.rst [hunk 1]
```diff
@@ -93,19 +93,71 @@ For the data types above:
...
-4. Cursor Tracking
+- Check that sliceviewer gets the projection correct for an ``MDHisto`` object from a non-axis aligned cut.
+
+.. code-block:: python
+
+	ws = CreateMDWorkspace(Dimensions='3', Extents='-3,3,-3,3,-3,3',
+                       Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
+                       Frames='HKL,HKL,HKL',
+                       SplitInto='2', SplitThreshold='10')
+    expt_info = CreateSampleWorkspace()
```
Comment (DanielMurphy22) [inline dev-docs/source/Testing/SliceViewer/SliceViewer.rst:114]:
```suggestion
    ws = CreateMDWorkspace(Dimensions='3', Extents='-3,3,-3,3,-3,3',
                       Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                       Frames='HKL,HKL,HKL',
                       SplitInto='2', SplitThreshold='10')
    expt_info = CreateSampleWorkspace()
```
just need an unindent

File: dev-docs/source/Testing/SliceViewer/SliceViewer.rst [hunk 2]
```diff
@@ -93,19 +93,71 @@ For the data types above:
...
+######################
+
 - For the ``SXD_MD_nonortho`` workspace, the non-orthogonal view button (see below) should be enabled - clicking this should also turn on grid lines. When viewing the H and K axes you should see the gridlines are no longer perpendicular to each other.
 
 .. figure:: ../../../../docs/source/images/wb-sliceviewer51-nonorthobutton.png
    :class: screenshot
    :align: center
 
-4. Cursor Tracking
+- Check that sliceviewer gets the projection correct for an ``MDHisto`` object from a non-axis aligned cut.
```
Comment (DanielMurphy22) [inline dev-docs/source/Testing/SliceViewer/SliceViewer.rst:106]: Can you reword this one line to be less sciencey? I think it currently means nothing to most of our dev testers as what does the "correct projection" look like? The other lines are well explained!

**Review by DanielMurphy22 — CHANGES_REQUESTED** [review]
I tried to reformat the new section 4 with letters, so see what you think of my code snippet below. I was trying to make it more clear what lines belonged together! The new tests are really good, just a few small things on formatting and being less sciencey! 😆 

```
A. For the ``SXD_MD_nonortho`` workspace, the non-orthogonal view button (see below) should be enabled - clicking this should also turn on grid lines. When viewing the H and K axes you should see the gridlines are no longer perpendicular to each other.

.. figure:: ../../../../docs/source/images/wb-sliceviewer51-nonorthobutton.png
   :class: screenshot
   :align: center

B. Check that sliceviewer gets the projection correct for an ``MDHisto`` object from a non-axis aligned cut.

  .. code-block:: python
  
     ws = CreateMDWorkspace(Dimensions='3', Extents='-3,3,-3,3,-3,3',
                            Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                            Frames='HKL,HKL,HKL',
                            SplitInto='2', SplitThreshold='10')
     expt_info = CreateSampleWorkspace()
     ws.addExperimentInfo(expt_info)
     SetUB(ws, 1,1,2,90,90,120)
     BinMD(InputWorkspace=ws, AxisAligned=False,
           BasisVector0='[00L],r.l.u.,0,0,1',
           BasisVector1='[HH0],r.l.u.,1,1,0',
           BasisVector2='[-HH0],r.l.u.,-1,1,0',
           OutputExtents='-4,4,-4,4,-0.25,0.25',
           OutputBins='101,101,1', OutputWorkspace='ws_slice', NormalizeBasisVectors=False)

  - Open ``ws_slice`` in the sliceviewer and check that non-orthogonal view button is 
    enabled (as it has a valid UB matrix) - however when clicked the 
    gridlines are orthogonal  (in this case 110 is orthogonal to 001).

C. Check that changing the properties of the workspace that governs the support for the non-orthogonal transform closes the sliceviewer window.

  - Open ``ws`` from the above test script (in 4B) in the sliceviewer

  - Run ``ClearUB(ws)`` (it should close sliceviewer with warning ``property supports_nonorthogonal_axes is different...``

  - Instead of clearing the UB you could replace the workspace with a workspace of a different frame that doesn't support the 
    non-orthogonal view (e.g. QLab, as in the following script)

  .. code-block:: python

	  ws = CreateMDWorkspace(Dimensions='3', Extents='-6,6,-4,4,-0.5,0.5',
                            Names='Q1,Q2,Q3', Units='Ang-1,Ang-1,Ang-1',
                            Frames='QLab,QLab,QLab',
                            SplitInto='2', SplitThreshold='50')

D. Check that the non-orthogonal view is disabled for non-Q axes such as energy

  .. code-block:: python
  
    ws_4D = CreateMDWorkspace(Dimensions=4, Extents=[-1, 1, -1, 1, -1, 1, -1, 1], Names="E,H,K,L",
                              Frames='General Frame,HKL,HKL,HKL', Units='meV,r.l.u.,r.l.u.,r.l.u.')
    expt_info_4D = CreateSampleWorkspace()
    ws_4D.addExperimentInfo(expt_info_4D)
    SetUB(ws_4D, 1, 1, 2, 90, 90, 120)

  When the Energy axis is viewed the non-orthogonal view should disabled. The button should be re-enabled when you view two Q-axes e.g. H and K.
```

File: dev-docs/source/Testing/SliceViewer/SliceViewer.rst (hunk 2 above)
Comment (RichardWaiteSTFC) [reply dev-docs/source/Testing/SliceViewer/SliceViewer.rst:106]: Ah I didn't intend that as an instruction itself, the "test" for getting the correct projection matrix/basis vectors as columns from the workspace is outlined in more detail in the lines below. I will clarify this and replace projection with basis vectors which is a term more people will be familiar with.

File: dev-docs/source/Testing/SliceViewer/SliceViewer.rst [hunk 3]
```diff
@@ -93,19 +93,71 @@ For the data types above:
...
+        OutputExtents='-4,4,-4,4,-0.25,0.25',
+        OutputBins='101,101,1', OutputWorkspace='ws_slice', NormalizeBasisVectors=False)
+
+- Run the above code and open ``ws_slice`` in the sliceviewer.
+- Check that non-orthogonal view is enabled on opening - however when clicked the gridlines are orthogonal  (in this case 110 is orthogonal to 001).
+
+C. Check that changing the properties of the workspace that governs the support for the non-orthogonal transform closes the sliceviewer window.
+- Open ``ws`` from the above test script in the sliceviewer
+
+- Run ``ClearUB(ws)`` (it should close sliceviewer with warning ``property supports_nonorthogonal_axes is different...``
```
Comment (DanielMurphy22) [inline dev-docs/source/Testing/SliceViewer/SliceViewer.rst:130]:
```suggestion
C. Check that changing the properties of the workspace that governs the support for the non-orthogonal transform closes the sliceviewer window.

- Open ``ws`` from the above test script in the sliceviewer

- Run ``ClearUB(ws)`` (it should close sliceviewer with warning ``property supports_nonorthogonal_axes is different...``)
```

Sorry, so very close, but I need a newline before the "Open ws" bullet point and an end bracket.

Comment (RichardWaiteSTFC) [reply dev-docs/source/Testing/SliceViewer/SliceViewer.rst:130]: OK thanks - will do this and rebase -i to rejig history a bit

Comment (DanielMurphy22) [reply dev-docs/source/Testing/SliceViewer/SliceViewer.rst]: I'm all happy with the docs now so I'll leave the full review to Martyn!

**Review by martyngigg — APPROVED** [review]
Looks great to me now!

**Comment by RichardWaiteSTFC** [comment]
Thanks both!

**Review by DanielMurphy22 — APPROVED** [review]

**Pull request merged**